### DISCUSSION TITLE
Is there any way to identify phrasal verbs in a lengthy text by using the compromise library?

### DISCUSSION THREAD

**Discussion by farhadibehnamdev** [nlp-compromise/nlp_compromise#1159] [Q&A]
I’m working on a text analysis project and need to identify phrasal verbs in various forms within a lengthy text. I’m particularly interested in understanding how to detect:

> Separable Phrasal Verbs - These can be split by inserting an object, e.g., “turn off” as in “Turn off the lights” or “Turn the lights off.”
> 
> Inseparable Phrasal Verbs - These cannot be split, e.g., “look after” as in “She will look after the children.”
> 
> Three-Word Phrasal Verbs - Consisting of a verb and two particles, e.g., “catch up with.”
> 
> Phrasal Verbs in Different Tenses - E.g., “give up” can appear as “give up,” “gave up,” or “will give up.”

What are some effective methods or tools for automatically identifying these types of phrasal verbs in text? I have found some discussions on phrasal verbs, but they are outdated and lack practical solutions.

**Comment by spencermountain** [comment]
hey Behnam, yep we will tag standard phrasal verbs, which are mostly defined [here](https://github.com/spencermountain/compromise/blob/master/data/lexicon/verbs/phrasals.js)
```js
doc.match('#PhrasalVerb+').json()
```
if you wanted to add chosen words on each side, I would do something like:
```js
doc.match('will? #PhrasalVerb+ (with|for|by)?').forEach(m=>{
   //add some kind of checks for your analysis
   if(m.after('^the').found){ }
   if(m.sentence().has('look after the')){    }
})
```
and you should be able to do any kind of further classification of them.
cheers

**Reply by farhadibehnamdev** [reply]
Thank you so much , the code below works almost correctly, but there are some cases where it produces incorrect outputs. For example, in English, we don’t use phrases like "documents up" or “would up” or "couldn't off" or "up over" yet these are output as phrasal verbs. I understand that “pick the documents up” is a correct phrase, but “documents up” is not. Similarly, in the case of “couldn't put off” the output correctly identifies “put off” as a phrasal verb, but it also incorrectly includes “couldn't off,” which is not accurate. another one "piled up over" however the "piled up" is correct and outputed, but also outputed "up over" that is not correct. How to increase accuracy to finding any phrasal verbs?
@spencermountain 

  ```
findPhrasalVerbs(text: string): string[] {
    const doc = nlp(text)
    const foundPhrasalVerbs = new Set<string>()

    const particles = [
      'up',
      'off',
      'out',
      'away',
      'in',
      'on',
      'for',
      'by',
      'with',
      'down',
      'over',
      'about',
      'back',
      'around',
      'through',
      'along',
      'across',
      'into',
      'onto',
      'toward',
      'from',
      'beneath',
      'under',
      'above',
      'beside',
      'among',
      'amongst',
      'beyond',
      'except',
      'since',
      'than',
      'via',
      'amid',
      'below',
      'circa',
      'despite',
      'including',
      'like',
      'minus',
      'plus',
      'versus',
      'within',
    ]

    const auxiliaryVerbs = new Set([
      'am',
      'is',
      'are',
      'was',
      'were',
      'be',
      'been',
      'being',
      'have',
      'has',
      'had',
      'having',
      'do',
      'does',
      'did',
      'doing',
     
    ])

    particles.forEach((particle) => {
      // Pattern 1: Direct phrasal verbs (e.g., "turn on")
      const compactPattern = `#Verb ${particle}`
      doc.match(compactPattern).forEach((match) => {
        const matchText = match.text()
        const cleanWord = matchText.replace(/-/g, '').trim()
        const [verb, matchedParticle] = cleanWord.split(' ')

        // Check if the verb is not an auxiliary verb
        if (
          matchedParticle.toLowerCase() === particle &&
          !auxiliaryVerbs.has(verb.toLowerCase())
        ) {
          foundPhrasalVerbs.add(cleanWord)
        }
      })

      // Pattern 2: Split phrasal verbs with up to 3 words in between
      const adjunctPattern = `#Verb .{0,3} ${particle}`
      doc.match(adjunctPattern).forEach((match) => {
        const matchText = match.text()
        const cleanWord = matchText.replace(/-/g, '').trim()
        const words = cleanWord.split(' ')

        const verb = words[0].replace(/-/g, '').trim()
        const lastWord = words[words.length - 1].toLowerCase()

        // Check if the verb is not an auxiliary verb and the last word is the particle
        if (lastWord === particle && !auxiliaryVerbs.has(verb.toLowerCase())) {
          foundPhrasalVerbs.add(`${verb} ${particle}`)
        }
      })
    })

    return Array.from(foundPhrasalVerbs)
  }
```

**Comment by spencermountain** [comment]
hey, thank you Behnam - it would be really useful to find situations where our phrasal-verb tagger is wrong. I think this is a really important part of sentence understanding, and I'm happy to fix any examples you can find.

I couldn't find any mistaken `PhrasalVerb` tags in the examples you mentioned. Can you help me reproduce them?

```js
txt = `i sent the documents up the hill`
txt = `he would up stage his friend`
txt = `he couldn't off gas`
txt = `he got up over the hill`
nlp(txt).debug()
```
If it were possible to get a list of sentences that have a mis-tagged phrasal, I'll add them to the tests.
thanks for your help
cheers
